topdown/tokens_verify.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// Implements RS256 JWT signature verification
+func builtinJWTVerifyRS256(a ast.Value, b ast.Value) (ast.Value, error) {
+
+	// Process the token
+	astToken, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+	token := string(astToken)
+	if !strings.Contains(token, ".") {
+		return nil, errors.New("encoded JWT has no period separators")
+	}
+
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("encoded JWT must have 3 sections, found %d", len(parts))
+	}
+
+	result := make([][]byte, len(parts))
+	for i, part := range parts {
+		if result[i], err = decode(part); err != nil {
+			return nil, err
+		}
+	}
+
+	header, payload, signature := result[0], result[1], result[2]
+	headerPayload := []byte(serialize(header, payload))
+
+	// Process PEM encoded certificate input
+	astCertificate, err := builtins.StringOperand(b, 2)
+	if err != nil {
+		return nil, err
+	}
+	certificate := string(astCertificate)
+
+	block, rest := pem.Decode([]byte(certificate))
+
+	if block == nil || block.Type != "CERTIFICATE" || len(rest) > 0 {
+		return nil, fmt.Errorf("failed to decode PEM block containing certificate")
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse PEM certificate %v", err)
+	}
+
+	// Get public key
+	publicKey := cert.PublicKey.(*rsa.PublicKey)
+
+	// Validate the JWT signature
+	err = rsa.VerifyPKCS1v15(publicKey, crypto.SHA256, getInputSHA(headerPayload), []byte(signature))
+
+	if err != nil {
+		return ast.String("RS256 JWT signature verification failed"), nil
+	}
+	return ast.String(""), nil
+
+}
+
+// getInputSha returns the SHA256 checksum of the input
+func getInputSHA(input []byte) (hash []byte) {
+	hasher := sha256.New()
+	hasher.Write(input)
+	return hasher.Sum(nil)
+}
+
+// decode decodes base64url string to byte array
+func decode(data string) ([]byte, error) {
+	data = strings.Replace(data, "-", "+", -1) // 62nd char of encoding
+	data = strings.Replace(data, "_", "/", -1) // 63rd char of encoding
+
+	switch len(data) % 4 { // Pad with trailing '='s
+	case 0: // no padding
+	case 2:
+		data += "==" // 2 pad chars
+	case 3:
+		data += "=" // 1 pad char
+	}
+
+	return base64.StdEncoding.DecodeString(data)
+}
+
+// encode encodes given byte array to base64url string
```

Comment:
Can this be replaced with `base64.URLEncoding.EncodeToString`?

Reply:
Yes it can. Do we want to ?

Reply:
Yes, let's use the standard library implementation instead.